lib/TimezoneOffsetUtil.js

```diff
@@ -0,0 +1,202 @@
+/*
+ * == BSD2 LICENSE ==
+ * Copyright (c) 2014, Tidepool Project
+ * 
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the associated License, which is identical to the BSD 2-Clause
+ * License as published by the Open Source Initiative at opensource.org.
+ * 
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the License for more details.
+ * 
+ * You should have received a copy of the License along with this program; if
+ * not, you can obtain one from Tidepool Project at tidepool.org.
+ * == BSD2 LICENSE ==
+ */
+
+'use strict';
+
+var _ = require('lodash');
+var util = require('util');
+
+var sundial = require('sundial');
+
+var annotate = require('./eventAnnotations');
+
+var isBrowser = typeof window !== 'undefined';
+var debug = isBrowser ? require('./bows')('TimezoneOffsetUtil') : console.log;
+
+module.exports = function(timezone, mostRecent, changes) {
+  var self = this;
+  self.type = 'across-the-board-timezone';
+
+  var offsetIntervals = [];
+  var currentOffset = null, currentIndex = null;
+
+  this.findOffsetDifference = function(event) {
+    return Math.round(
+      // some timezones have offsets in units of 15 minutes (e.g., +12:45)
+      // so we round to the nearest 15 minutes whenever we see an offset change
+      sundial.dateDifference(event.change.from, event.change.to, 'minutes')/15
+    ) * 15;
+  };
+
+  function adjustCurrentOffset(event) {
+    var difference = self.findOffsetDifference(event);
+    // the farthest timezones are UTC+14 and UTC-12, so max offset
+    // difference is 840 + 720 = 1560
+    var MAX_DIFF = 1560;
+    if (Math.abs(difference) <= MAX_DIFF) {
+      currentOffset += difference;
+    }
+  }
+
+  // check that an appropriate timezone is passed in
+  try {
+    sundial.checkTimezoneName(timezone);
+  }
+  catch (e) {
+    throw e;
+  }
+
+  // check that an appropriate date of most recent datum is passed in
+  if (isNaN(Date.parse(mostRecent))) {
+    throw new Error('Invalid timestamp for most recent datum!');
+  }
+
+  // check that the changes are of the correct type if passed in
+  if (!_.isEmpty(changes)) {
+    _.each(changes, function(change) {
+      if (!(change.subType && change.subType === 'timeChange')) {
+        var error = new Error(util.format('Wrong subType of object passed as `timeChange`. Object is [%s]', JSON.stringify(change)));
+        throw error;
+      }
+      self.type = 'utc-bootstrapping';
```

Comment:
Setting the type doesn't have to be done inside the _.each loop, right?  It makes it look like it's part of the 'if', or like it depends somehow on the nature of one of the 'changes' objects